src/fsharp/FSharp.Build/FSharpEmbedResXSource.fs

```diff
@@ -0,0 +1,144 @@
+                        |> Array.fold (fun (sb:StringBuilder) line -> sb.AppendLine("    /// " + line)) (StringBuilder())
+                    // add the resource
+                    let accessorBody =
+                        match (generateLegacy, generateLiteral) with
+                        | (true, true) -> sprintf "    [<Literal>]\n    let %s = \"%s\"" identifier name
+                        | (true, false) -> sprintf "    let %s = \"%s\"" identifier name // the [<Literal>] attribute can't be used for FSharp.Core
+                        | (false, _) ->
+                            let isStringResource = match node.Attribute(xname "type") with
+                                                   | null -> true
+                                                   | _ -> false
+                            match (isStringResource, generateGetObject) with
+                            | (true, _) -> sprintf "    let %s() = GetString(\"%s\")" identifier name
+                            | (false, true) -> sprintf "    let %s() = GetObject(\"%s\")" identifier name
+                            | (false, false) -> "" // the target runtime doesn't support non-string resources
+                            // TODO: When calling the `GetObject` version, parse the `type` attribute to discover the proper return type
+                    sb.AppendLine().Append(commentBody).AppendLine(accessorBody)
+                ) sb
+            File.WriteAllText(sourcePath, body.ToString())
+            printMessage <| sprintf "Done: %s" sourcePath
+            Some(sourcePath)
+        with e ->
+            printf "An exception occurred when processing '%s'\n%s" resx (e.ToString())
+            None
+
+    [<Required>]
+    member this.EmbeddedResource
+        with get() = _embeddedText
+         and set(value) = _embeddedText <- value
+
+    [<Required>]
+    member this.IntermediateOutputPath
+        with get() = _outputPath
+         and set(value) = _outputPath <- value
+
+    member this.TargetFramework
+        with get() = _targetFramework
+         and set(value) = _targetFramework <- value
+
+    [<Output>]
+    member this.GeneratedSource
+        with get() = _generatedSource
+
+    interface ITask with
+        member this.BuildEngine
+            with get() = _buildEngine
+             and set(value) = _buildEngine <- value
+        member this.HostObject
+            with get() = _hostObject
+             and set(value) = _hostObject <- value
+        member this.Execute() =
+            let getBooleanMetadata (metadataName:string) (defaultValue:bool) (item:ITaskItem) =
+                match item.GetMetadata(metadataName) with
+                | value when String.IsNullOrWhiteSpace(value) -> defaultValue
+                | value -> String.Compare(value, "true", StringComparison.OrdinalIgnoreCase) = 0
+            let generatedFiles, generatedResult =
```

Comment:
Please check the latest code in https://github.com/Microsoft/visualfsharp/pull/3614/files#diff-f583ab0cc0cd86307b6b05f78682f65a for a cleaner way to write this. A general rule is not to use ``fold`` unless you have a truly generative process where the results from one step depend on the results from the previous step

Reply:
Fixed.  Extra commits coming soon.